[2.x] Allow automatic seeding after automatic migrations

drbyte opened the pull request:
Fixes #158

If `config('tenancy.seeder_class')` is null or a blank string, the seeder-class parameter is not passed, and therefore Laravel's default option of calling 'DatabaseSeeder' will occur.

As with Laravel's normal seeding process, the class is treated as the "root" seeder class, which can call other seeder classes in turn.

stancl commented on src/TenantManager.php:
```diff
@@ -71,6 +71,14 @@ public function createTenant(Tenant $tenant): self
+                    $this->artisan->call('tenants:seed', [
+                        '--tenants' => [$tenant['id']],
+                    ] + $seederClassParameter);
+                }
```
```php
if ($this->shouldSeedAfterMigration()) {
    $afterCreating += $this->databaseCreationQueued() ? [
        new QueuedTenantDatabaseSeeder($tenant),
    ] : [
        function () use ($tenant) {
            $this->artisan->call('tenants:seed', [
                '--tenants' => [$tenant['id']],
            ]); // + the options
        },
    ];
}
```

stancl commented on src/TenantManager.php:
```diff
@@ -311,6 +319,11 @@ public function shouldMigrateAfterCreation(): bool
 
+    public function shouldSeedAfterMigration(): bool
+    {
+        return $this->app['config']['tenancy.seed_after_migration'] ?? false;
```
Safety check:

```suggestion
        return $this->shouldMigrateAfterCreation() && $this->app['config']['tenancy.seed_after_migration'] ?? false;
```

stancl commented on src/Jobs/QueuedTenantDatabaseMigrator.php:
```diff
@@ -34,5 +42,11 @@ public function handle()
             '--tenants' => [$this->tenant->id],
         ]);
+
+        if ($this->shouldRunSeeder) {
```
Could you extract this logic into a separate, queuable `QueuedTenantDatabaseSeeder`?

stancl reviewed (commented):
I made some changes to how the queued jobs work:
- there's now only one config that decides if the jobs will be queued (`queue_database_creation`)
- job chains are used to prevent race conditions

Could you fix conflicts & update some things as per my review?

If you don't want to deal with the conflicts feel free to create a new PR or I can create a new PR.

drbyte committed:
[2.1.0] Allow automatic seeding after automatic migrations

Fixes #158

If config('tenancy.seeder_class') is null or a blank string, the seeder-class parameter is not passed, and therefore Laravel's default option of calling 'DatabaseSeeder' will occur.

As with Laravel's normal seeding process, the class is treated as the "root" seeder class, which can call other seeder classes in turn.

drbyte committed:
Add auto-seeder tests

drbyte commented on src/TenantManager.php:
```diff
@@ -67,15 +68,30 @@ public function createTenant(Tenant $tenant): self
 
         if ($this->shouldMigrateAfterCreation()) {
-            $afterCreating += $this->databaseCreationQueued() ? [
+            $afterCreating = array_merge($afterCreating, $this->databaseCreationQueued() ? [
```
Note: I changed these to `array_merge` because the `+=` was behaving like `array_unique` instead. :(

stancl replied:
Yeah, I also noticed some weird behavior with that operator. Thought the syntax looked nice. Look at my PR to your `seeder` branch, I use `[]` there, since the arrays have only one item anyway.

drbyte replied:
Oh, duh. Just adding using `$foo[] =` is definitely easier!

stancl committed:
Refine seeding tests (#1)

Refine seeding tests

drbyte commented on tests/TenantManagerTest.php:
```diff
@@ -235,6 +237,28 @@ public function automatic_migrations_work()
         $tenant2 = Tenant::create(['bar.localhost']);
         tenancy()->initialize($tenant2);
         $this->assertTrue(\Schema::hasTable('users'));
+        $this->assertCount(0, \DB::select('select * from users'));
```
I guess this could be changed to `assertSame` too?
Maybe it's not even needed.
I only included it to differentiate between outcomes of migrations vs seeders.

stancl replied:
Oh, yeah, nice catch.

drbyte replied:
Maybe it's not even needed?

stancl replied:
Actually, on second thought, I'd remove the assertion. Just assert that the table doesn't exist. Making queries to nonexistent tables is confusing and I'm surprised it doesn't throw an exception.

drbyte replied:
Removal makes sense.

There's no exception because the table does exist at that point. Just empty

stancl replied:
Ah, I misread the line above, thought it's `assertFalse`. But yeah, I'd remove it anyway because we're testing migrations at this point.

drbyte replied:
Done

drbyte committed:
Update TenantManagerTest.php

Remove unnecessary assertion.
